### PR TITLE
Allow multiline prompts

### LINKED ISSUE #1275: LitGPT chat terminates weirdly
Our `litgpt chat` command is doing something weird where it exits the shell and executes the code on the terminal.

To reproduce:

```
litgpt chat   --checkpoint_dir out/custom-phi-2/final
```

and then paste the following into the chat window:

```
explain this code: # 1) Download a pretrained model
litgpt download --repo_id microsoft/phi-2

# 2) Pretrain the model
litgpt pretrain \
  --initial_checkpoint_dir checkpoints/microsoft/phi-2 \
  --data Alpaca2k \
  --out_dir out/custom-phi-2

# 3) Chat with the model
litgpt chat \
  --checkpoint_dir out/custom-phi-2/final
```

This will first work correctly but then exit the session and run the code on the shell, e.g. search for `>> Prompt: >> Reply` in the output below:

```
⚡ codegemma ~/litgpt litgpt chat --checkpoint_dir checkpoints/google/codegemma-7b-it
Now chatting with CodeGemma-7b-it.
To exit, press 'Enter' on an empty prompt.

Seed set to 1234
>> Prompt: explain this code: # 1) Download a pretrained model
litgpt download --repo_id microsoft/phi-2

# 2) Pretrain the model
litgpt pretrain \
  --initial_checkpoint_dir checkpoints/microsoft/phi-2 \
  --data Alpaca2k \
  --out_dir out/custom-phi-2

# 3) Chat with the model
litgpt chat \
  --checkpoint_dir out/custom-phi-2/final


>> Reply: The code snippet you provided attempts to download a pre-trained model from a given URL. Here's a step-by-step explanation:

1. **Import the necessary library:** The code starts by importing the `requests` library, which is needed to make HTTP requests to the URL.

2. **Define the URL:** The URL where the pre-trained model is located is defined as a string.

3. **Download the model:** The `requests.get()` method is used to send a GET request to the URL. This request retrieves the binary data of the pre-trained model.

4. **Save the model:** The `with open()` statement is used to open a file named `model.ckpt` in write binary mode (`wb`). The retrieved model data is then written to this file using the file object's `write()` method.

5. **Close the file:** The file object is closed using the `close()` method to save the changes made to the file.

In summary, this code downloads a pre-trained model from the specified URL and saves it locally as a file named `model.ckpt`. This can be useful for using the model in further applications or training additional models.
Time for inference: 13.50 sec total, 18.52 tokens/sec, 250 tokens

>> Prompt: >> Reply: I can't download the LITGPT model because it is not publicly available. It is a private repository on the Hugging Face model hub. For security reasons, I can't access private repositories without the owner's permission.
Time for inference: 2.47 sec total, 19.00 tokens/sec, 47 tokens

>> Prompt: %                                                                                                                    
⚡ codegemma ~/litgpt # 2) Pretrain the model
⚡ codegemma ~/litgpt litgpt pretrain \
>   --initial_checkpoint_dir checkpoints/microsoft/phi-2 \
>   --data Alpaca2k \
>   --out_dir out/custom-phi-2
Traceback (most recent call last):
  File "/home/zeus/miniconda3/envs/cloudspace/bin/litgpt", line 8, in <module>
    sys.exit(main())
  File "/teamspace/studios/this_studio/litgpt/litgpt/__main__.py", line 131, in main
    fn(**kwargs)
  File "/teamspace/studios/this_studio/litgpt/litgpt/pretrain.py", line 96, in setup
    raise ValueError(f"Please specify --model_name <model_name>. Available values:\n{available_models}")
ValueError: Please specify --model_name <model_name>. Available values:
Camel-Platypus2-13B
Camel-Platypus2-70B
CodeGemma-7b-it
CodeLlama-13b-Instruct-hf
CodeLlama-13b-Python-hf
CodeLlama-13b-hf
CodeLlama-34b-Instruct-hf
CodeLlama-34b-Python-hf
CodeLlama-34b-hf
CodeLlama-70b-Instruct-hf
CodeLlama-70b-Python-hf
CodeLlama-70b-hf
CodeLlama-7b-Instruct-hf
CodeLlama-7b-Python-hf
CodeLlama-7b-hf
FreeWilly2
Gemma-2b
Gemma-2b-it
Gemma-7b
Gemma-7b-it
LLaMA-2-7B-32K
Llama-2-13b-chat-hf
Llama-2-13b-hf
Llama-2-70b-chat-hf
Llama-2-70b-hf
Llama-2-7b-chat-hf
Llama-2-7b-chat-hf-function-calling-v2
Llama-2-7b-hf
Mistral-7B-Instruct-v0.1
Mistral-7B-Instruct-v0.2
Mistral-7B-v0.1
Mistral-7B-v0.2
Mixtral-8x7B-Instruct-v0.1
Mixtral-8x7B-v0.1
Nous-Hermes-13b
Nous-Hermes-Llama2-13b
Nous-Hermes-llama-2-7b
Platypus-30B
Platypus2-13B
Platypus2-70B
Platypus2-70B-instruct
Platypus2-7B
RedPajama-INCITE-7B-Base
RedPajama-INCITE-7B-Chat
RedPajama-INCITE-7B-Instruct
RedPajama-INCITE-Base-3B-v1
RedPajama-INCITE-Base-7B-v0.1
RedPajama-INCITE-Chat-3B-v1
RedPajama-INCITE-Chat-7B-v0.1
RedPajama-INCITE-Instruct-3B-v1
RedPajama-INCITE-Instruct-7B-v0.1
Stable-Platypus2-13B
dolly-v2-12b
dolly-v2-3b
dolly-v2-7b
falcon-180B
falcon-180B-chat
falcon-40b
falcon-40b-instruct
falcon-7b
falcon-7b-instruct
longchat-13b-16k
longchat-7b-16k
open_llama_13b
open_llama_3b
open_llama_7b
phi-1_5
phi-2
pythia-1.4b
pythia-1.4b-deduped
pythia-12b
pythia-12b-deduped
pythia-14m
pythia-160m
pythia-160m-deduped
pythia-1b
pythia-1b-deduped
pythia-2.8b
pythia-2.8b-deduped
pythia-31m
pythia-410m
pythia-410m-deduped
pythia-6.9b
pythia-6.9b-deduped
pythia-70m
pythia-70m-deduped
stable-code-3b
stablecode-completion-alpha-3b
stablecode-completion-alpha-3b-4k
stablecode-instruct-alpha-3b
stablelm-3b-4e1t
stablelm-base-alpha-3b
stablelm-base-alpha-7b
stablelm-tuned-alpha-3b
stablelm-tuned-alpha-7b
stablelm-zephyr-3b
tiny-llama-1.1b
tiny-llama-1.1b-chat
vicuna-13b-v1.3
vicuna-13b-v1.5
vicuna-13b-v1.5-16k
vicuna-33b-v1.3
vicuna-7b-v1.3
vicuna-7b-v1.5
vicuna-7b-v1.5-16k
⚡ codegemma ~/litgpt
⚡ codegemma ~/litgpt # 3) Chat with the model
⚡ codegemma ~/litgpt litgpt chat \
>   --checkpoint_dir out/custom-phi-2/final
--checkpoint_dir '/teamspace/studios/this_studio/litgpt/out/custom-phi-2/final' is not a checkpoint directory.
Find download instructions at https://github.com/Lightning-AI/litgpt/blob/main/tutorials

You have downloaded locally:
 --checkpoint_dir '/teamspace/studios/this_studio/litgpt/checkpoints/google/codegemma-7b-it'

See all download options by running:

```

### PR DESCRIPTION
**Pull request by rasbt**
This workaround let's users submit multiline prompts. Either via copy & paste or by typing on the command line and using Enter once to start a new line. If the newline is blank and a user hits Enter again, it submits the prompt. 

If a user only wants to submit a single line prompt, pressing Enter twice might be extra effort, but I think it's worth it given that users can now use multiline prompts without "bugs". 

We can also explore other options based on external libraries some time later but I think it's kind of useful to fix this so users can use CodeGemma etc.

Fixes #1275

### CODE REVIEW

File: litgpt/chat/base.py [hunk 1]
```diff
@@ -163,14 +174,18 @@ def main(
     )
     stop_tokens = prompt_style.stop_tokens(tokenizer)
 
-    print(f"Now chatting with {config.name}.\nTo exit, press 'Enter' on an empty prompt.\n")
+    print(
+        f"Now chatting with {config.name}.\nTo exit, type 'exit'"
+        " or 'quit' and press 'Enter'.\nAfter entering your prompt,"
+        " hit 'Enter' once to start writing on a new line. Hit 'Enter'"
+        " twice to submit the prompt.")
```
Comment (carmocca) [inline litgpt/chat/base.py:181]: I find it a bit arbitrary to support 1 line break but not 2, or 2 but not 3, ...

It would be worth to instead go for a more complete solution directly.

For instance, https://github.com/simonw/llm makes you enclose your multi-line test in a 

```
!multi
your text
...
...
!end
```

block.


Another alternative would be to find out if we register Ctrl+Enter to send the prompt so that Enter can be a simple line break.

There might be other solutions

**Comment by awaelchli** [comment]
I'm good with whatever the solution if it does not impact the single-line user experience. 

However, keep in mind that if you add special syntax for the multi-line input, it won't solve your original issue https://github.com/Lightning-AI/litgpt/issues/1275 where you copy-pasted multi-line into the terminal.

**Comment by rasbt** [comment]
Ok, thinking about this more, I think there is no one-size-fits-all approach. 

So, I changed it back so that by default nothing changes. 

But you can now additionally pass the `--multiline_prompts` arg:

```
litgpt chat \
   --checkpoint_dir checkpoints/microsoft/phi-2 \
   --multiline_prompts True
```

to work with multiline prompts. This will then require users to enter `!submit` to submit. (I found that `:submit:` is just too annoying to type on US keyboards.)

This should also address #1275 based on trying it.

The code should be ready for review @carmocca and @awaelchli .

File: litgpt/chat/base.py [hunk 2]
```diff
@@ -110,6 +110,51 @@ def decode(fabric: L.Fabric, tokenizer: Tokenizer, token_stream: Iterator[torch.
     return tokens_generated
 
 
+def process_prompt(prompt, model, tokenizer, prompt_style, fabric, temperature, top_k, top_p, stop_tokens):
+    if not prompt.lower().strip() in ("", "quit", "exit"):
```
Comment (carmocca) [inline litgpt/chat/base.py:114]: Why is this necessary if you are also checking it in L152?

If it's really necessary, I would add the nit that it should do an early return to avoid the extra indentation level for the whole code block

File: litgpt/chat/base.py [hunk 3]
```diff
@@ -110,6 +110,51 @@ def decode(fabric: L.Fabric, tokenizer: Tokenizer, token_stream: Iterator[torch.
...
+                    line = input()
+                    if line.strip().lower() == "!submit":
+                        break
+                    prompt_lines.append(line)
+                prompt = "\n".join(prompt_lines)
+
+        except KeyboardInterrupt:
+            break
+
+        if prompt.lower().strip() in ("quit", "exit"):
```
Comment (carmocca) [inline litgpt/chat/base.py:152]: Should we have !quit and !exit now?

**Review by carmocca — COMMENTED** [review]
Some tests are failing

File: litgpt/chat/base.py (hunk 2 above)
Comment (rasbt) [reply litgpt/chat/base.py]: Good callout. I cleaned this up.

File: litgpt/chat/base.py (hunk 3 above)
Comment (rasbt) [reply litgpt/chat/base.py]: I think this makes sense for consistency. I updated it.

**Comment by rasbt** [comment]
All the feedback should be addressed now. Thanks @carmocca @awaelchli for keeping the bar high!

The default usage is the same as before, except that you can now exit via `!exit` and `!quit` in addition to `'Enter'` on a blank line:

```
⚡ multiline ~/litgpt litgpt chat --checkpoint_dir checkpoints/microsoft/phi-2
Now chatting with phi-2.
To exit, press 'Enter' on an empty prompt.

Seed set to 1234
>> Prompt: what do llamas eat/
>> Reply:  Llamas eat grass, shrubs, and leaves.

Time for inference: 0.74 sec total, 17.47 tokens/sec, 13 tokens

>> Prompt: !exit
⚡ multiline ~/litgpt 
```

And with `--multiline true`, you exit via `!exit` and `!submit` as well:

```
>> Prompt: (Type '!submit' on a new line to end input).
!exit
⚡ multiline ~/litgpt litgpt chat --checkpoint_dir checkpoints/microsoft/phi-2 --multiline true
Now chatting with phi-2.
To exit, enter '!quit' or '!exit' on an empty prompt and press 'Enter'.

Seed set to 1234
>> Prompt: (Type '!submit' on a new line to end input).
Hi
I have a question:
What do Llamas eat?
!submit
>> Reply:  llamas eat grass and hay.

Time for inference: 0.75 sec total, 10.64 tokens/sec, 8 tokens

>> Prompt: (Type '!submit' on a new line to end input).
!exit
```

File: litgpt/chat/base.py [hunk 4]
```diff
@@ -120,6 +165,7 @@ def main(
     quantize: Optional[Literal["bnb.nf4", "bnb.nf4-dq", "bnb.fp4", "bnb.fp4-dq", "bnb.int8"]] = None,
     precision: Optional[str] = None,
     compile: bool = False,
+    multiline_prompts: bool = False,
```
Comment (carmocca) [inline litgpt/chat/base.py:168]: What else could be multiline?

```suggestion
    multiline: bool = False,
```

**Review by carmocca — APPROVED** [review]

File: litgpt/chat/base.py (hunk 4 above)
Comment (rasbt) [reply litgpt/chat/base.py:168]: agreed, let's simplify
